zerver/data_import/mattermost.py

```diff
@@ -0,0 +1,540 @@
+        timezone=timezone,
+    )
+
+def convert_user_data(user_handler: UserHandler,
+                      user_id_mapper: IdMapper,
+                      user_data_map: Dict[str, Dict[str, Any]],
+                      realm_id: int,
+                      team_name: str) -> None:
+
+    user_data_list = []
+    for username in user_data_map:
+        user = user_data_map[username]
+        if check_user_in_team(user, team_name) or user["is_mirror_dummy"]:
+            user_data_list.append(user)
+
+    for raw_item in user_data_list:
+        user = process_user(raw_item, realm_id, team_name, user_handler, user_id_mapper)
+        user_handler.add_user(user)
+
+def convert_channel_data(channel_data: List[ZerverFieldsT],
+                         user_data_map: Dict[str, Dict[str, Any]],
+                         subscriber_handler: SubscriberHandler,
+                         stream_id_mapper: IdMapper,
+                         user_id_mapper: IdMapper,
+                         realm_id: int,
+                         team_name: str) -> List[ZerverFieldsT]:
+    channel_data_list = [
+        d
+        for d in channel_data
+        if d['team'] == team_name
+    ]
+
+    channel_members_map = {}  # type: Dict[str, List[str]]
+    channel_admins_map = {}  # type: Dict[str, List[str]]
+
+    def initialize_stream_membership_dicts() -> None:
+        for channel in channel_data:
+            channel_name = channel["name"]
+            channel_members_map[channel_name] = []
+            channel_admins_map[channel_name] = []
+
+        for username in user_data_map:
+            user_dict = user_data_map[username]
+            teams = user_dict["teams"]
+            for team in teams:
+                if team["name"] != team_name:
+                    continue
+                for channel in team["channels"]:
+                    channel_roles = channel["roles"]
+                    channel_name = channel["name"]
+                    if "channel_admin" in channel_roles:
+                        channel_admins_map[channel_name].append(username)
+                    elif "channel_user" in channel_roles:
+                        channel_members_map[channel_name].append(username)
+
+    def get_invite_only(v: str) -> bool:
+        if v == 'O':
+            return False
+        elif v == 'P':
+            return True
+        else:
+            raise Exception('unexpected value')
+
+    streams = []
+    initialize_stream_membership_dicts()
+
+    for channel_dict in channel_data_list:
+        now = int(timezone_now().timestamp())
+        stream_id = stream_id_mapper.get(channel_dict['name'])
+        stream_name = channel_dict["name"]
+        invite_only = get_invite_only(channel_dict['type'])
+
+        stream = build_stream(
+            date_created=now,
+            realm_id=realm_id,
+            name=channel_dict['display_name'] or channel_dict['name'],
+            description=channel_dict["purpose"] or channel_dict['header'],
```

Comment:
I'd add a quick comment before this line explaining that our description aligns well with Mattermost's purpose, but is displayed in a place not too different from their header, and so we include either if it's set.  (Assuming that's accurate, I haven't actually played with these fields)

Reply:
Yup. Following are the definitions of the terminologies according to Mattermost.

> Purpose(optional) - Describe how this channel should be used.
Header(optional) - Set text that will appear in the header of the channel beside the channel name. For example, include frequently used links by typing [Link Title](http://example.com).